Adds postprocessor to track time step number for transient problems

anshchaube opened the pull request:
Minor quality-of-life improvement for users who want to track time step counts during their MOOSE simulations.

closes #29127

anshchaube committed:
Adds postprocessor to track time step number for transient problems

Minor quality-of-life improvement for users who want to track
time step counts during their MOOSE simulations.

closes #29127

GiudGiud commented on framework/doc/content/source/postprocessors/NumTimeSteps.md:
```diff
@@ -0,0 +1,17 @@
+# TimestepSize
```
```suggestion
# NumTimeSteps
```

GiudGiud commented on framework/doc/content/source/postprocessors/NumTimeSteps.md:
```diff
@@ -0,0 +1,17 @@
+
+!syntax description /Postprocessors/NumTimeSteps
+
+## Description
+
+`NumTimeSteps` reports the timestep size.
+
```
```suggestion
```

GiudGiud commented on framework/include/postprocessors/NumTimeSteps.h:
```diff
@@ -0,0 +1,31 @@
+
+#include "GeneralPostprocessor.h"
+
+class NumTimeSteps : public GeneralPostprocessor
```
```suggestion
/**
 * Reports on the number of time steps already taken in the transient
 */
class NumTimeSteps : public GeneralPostprocessor
```

GiudGiud commented on test/tests/postprocessors/num_time_steps/tests:
```diff
@@ -0,0 +1,9 @@
+    type = 'CSVDiff'
+    input = 'numtimesteps.i'
+    csvdiff = 'numtimesteps_out.csv'
+    requirement = 'The postprocessor `NumTimeSteps` shall count the number of time-steps accurately'
```
```suggestion
    requirement = 'The shall be able to count the number of time-steps that have been taken '
```

GiudGiud commented on test/tests/postprocessors/num_time_steps/tests:
```diff
@@ -0,0 +1,9 @@
+    input = 'numtimesteps.i'
+    csvdiff = 'numtimesteps_out.csv'
+    requirement = 'The postprocessor `NumTimeSteps` shall count the number of time-steps accurately'
+                   'over the course of a transient solve'
```
```suggestion
                   'over the course of the simulation of a transient.'
```

GiudGiud reviewed (changes requested):
nice this is almost ready

anshchaube committed:
address review
- updated wording, file names based on suggestions
- added docstrings
- simplified test, changed dt

GiudGiud commented:
let's squash all these commits togther in just a
```
address review
- updated wording
- added docstrings
```
commit message

loganharbour commented on test/tests/postprocessors/num_time_steps/moose.i:
```diff
@@ -0,0 +1,98 @@
+[Mesh]
```
There is no need for a complex solve to test this. The test should be as minimal as possible to satisfy the requirement listed in the test spec:

> The shall be able to count the number of time-steps that have been taken over the course of the simulation of a transient

A coupled heat conduction solve and a large 27k element mesh is absolutely overkill for testing this postprocessor.

A basic problem without a solve is sufficient:

```
[Mesh/gmg]
  type = GeneratedMeshGenerator
  dim = 1
[]

[Postprocessors/timestep_ctr]
  type = NumTimeSteps
[]

[Problem]
  solve = False
[]

[Executioner]
  type = Transient
  dt = 1
  num_steps = 10
[]

[Outputs]
  csv = true
[]
```

dschwen commented on test/tests/postprocessors/num_time_steps/moose.i:
```diff
@@ -0,0 +1,98 @@
+  type = Transient
+  nl_abs_tol = 1e-10
+  nl_rel_tol = 1e-4
+  dt = 1
```
I suggest setting `dt` to something that doesn't result the time being the exact same as the time step number...

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2561525/), step Docs: sync website on 6066950 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29128/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2569039/), step Generate coverage on 3c176f1 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>cedfd0</th>
      <th colspan=4>#29128 3c176f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.14%</td>
      <td>85.14%</td>
      <td>+0.00%</td>
      <td>91.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>107658</td>
      <td>107669</td>
      <td>+11</td>
      <td>11</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18792</td>
      <td>18793</td>
      <td>+1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29128/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29128/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29128/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud committed:
Add SQA items to test configuration

Apply suggestions from code review

corrects code listing path

GiudGiud commented on test/tests/postprocessors/num_time_steps/tests:
```diff
@@ -0,0 +1,11 @@
+[Tests]
+  issues = #29127
```
```suggestion
  issues = '#29127'
```

GiudGiud commented:
Test failures unrelated (fenix failures are known, unrelated tests w/ openmpi)
